### ISSUE TITLE
Last file in series of .jpg downloaded always seems to be corrupt 950byte file

### ISSUE THREAD

**Issue by stl3** [7USTIN/manga-cli#5]
Using Raspberry Pi4B+
OS: Debian GNU/Linux 11 (bullseye)
All dependencies for manga-cli present
**1st issue:**
manga-cli fetches links to images, downloads them, but fails to convert *.jpg to pdf, due to policy issue, which results in 0byte pdf file and the error "`Enter input: error: could not open document`".
Upon manually checking the command with --verbose, we can see where the issue is:
```
$convert -verbose /home/dietpi/.cache/manga-cli/Berserk/chapter_4/*.jpg blah.pdf
---snip some stuff for clarity---
convert-im6.q16: attempt to perform an operation not allowed by the security policy `PDF' @ error/constitute.c/IsCoderAuthorized/421.
```
Solution
https://cromwell-intl.com/open-source/pdf-not-authorized.html
`nano /etc/ImageMagick-6/policy.xml`
Change policy for:
`<policy domain="coder" rights="read|write" pattern="PDF,PS" />`
**2nd Issue:**
It seems that mostly the last .jpg downloaded is always the corrupt one, for me at least. 
Steps to produce:
Search for "Kuro no Shoukanshi"
Select 1st chapter
manga-cli downloads 36 jpg files, of which 36,jpg (950 bytes) is the corrupt culprit that prevents the convert to pdf from happening. In fact, it was happening to me in most manga I searched, Berserk was another one (91.jpg). It's always the last jpg in the series that causes the problem, being 950bytes. So I added:
`                find "${image_dir}" -name "*.jpg" -size -4k -delete`
before 
`convert -verbose "${image_dir}/*.jpg" "${pdf_dir}"` 
which fixed the problem for me
I also opted for using img2pdf instead of ImageMagicks 'convert' (no idea if there are caveats or if it was a bad idea overall) but on my pi it seemed faster at the conversion. The line I added instead of the above convert line is this:
`img2pdf --output "${pdf_dir}" ${image_dir}/*.jpg`

Overall I think the only main problem is the the last 950byte jpg file generated, the other things I faced aren't really major issues.

**Comment by SuperNici** [comment]
I tested the 2nd Issue and got very different results. 

https://user-images.githubusercontent.com/51826257/152747810-d89bd34c-a017-4a70-a7bb-05e078420255.mp4

Im somehow getting more images than those that are actually displayed on [that chapters page](https://readmanganato.com/manga-bn978948/chapter-1)?

**Comment by stl3** [comment]
Hm, actually now you mention it, I went and checked the page for this manga, and it seems the stuff we are seeing at the bottom from manga-cli should be at the top? The part that comes after "we need donations" - this is meant to be pages 4-9 but has somehow been appended to the bottom of the pdf in the conversion process lol.
PS - sorry I should've seperated the issues, will do so next time.

**Comment by SuperNici** [comment]
Thanks for the insight, I haven't read this manga to there's no way i could have known. Although I'm noticing that the order is messed up with nearly every manga I've tried, will open another issue on this.

**Comment by stl3** [comment]
I actually came across this one when I did a search for Berserk and just used it as an example because the first chapter was fast to load. I think perhaps `convert_imgs_to_pdf` needs something to guarantee the images are being fed sequentially (I don't actually know if convert does that by default).

**Comment by stl3** [comment]
Ok, after some testing, it seems that all we need to do is make sure there are leading 0's on the *.jpgs - so like 001.jpg, 002.jpg.
I tested both with imagemagick and img2pdf and the page sequences were correct but I use some copy pasta old code I found:
```
# Make pictures sequential
                find "${image_dir}" -type f -name '*.jpg' -exec bash -c '
                    for pathname do
                    dirpath=${pathname%/*}  # or: dirpath=$(dirname "$pathname")
                    name=${pathname##*/}    # or: name=$(basename "$pathname")
                        number=${name##*_}
                        number=${number%.jpg}
                    printf -v newname "%s_%.3d.jpg" seq "$number"
        mv "$pathname" "$dirpath/$newname"
    done' bash {} +

```
Only use this code for testing though, there are better ways to do this lol. Paste it above this line:
`convert "${image_dir}/*.jpg" "${pdf_dir}"`

I've only compared it against how the website was displaying this manga, but it looks right so far.

**Comment by 7USTIN** [comment]
Huge thanks, @stl and @supernici! Thanks for the great work!

I just pushed a commit that should fix this issue and issue #6.

The first problem should be fixed by changing ImageMagick's 'convert' command to img2pdf.

The second problem origined from also scraping javascript from the website and saving it as a jpg lol. The solution to this was to simply fix the scrape function.

Also a good call on img2pdf. It's much faster than ImageMagick's conversion command!

**Comment by SuperNici** [comment]
Works like a charm! Thanks for all the work you two!
